ada_feeding_perception/ada_feeding_perception/face_detection.py

```diff
@@ -0,0 +1,314 @@
+            v = int(self.img_mouth_center[1])
+
+            # Retrieve the depth value for this 2d coordinate
+            width = msg.width
+            depth = msg.data[(u % width) + (v * width)]
+
+            # Create target 3d point
+            mouth_location = PointStamped()
+            mouth_location.header = msg.header
+            mouth_location.point.x = float(u)
+            mouth_location.point.y = float(v)
+            mouth_location.point.z = float(depth)
+
+            # Publish 3d point
+            face_detection_msg = FaceDetection()
+            face_detection_msg.is_face_detected = self.is_face_detected
+            face_detection_msg.detected_mouth_center = mouth_location
+            self.publisher_results.publish(face_detection_msg)
+
+    def camera_callback(self, msg):
+        """
+        Callback function for the camera feed. If face detection is on, this
+        function will detect faces in the image and publish information about
+        them to the /face_detection topic.
+        """
+        self.is_on_lock.acquire()
+        is_on = self.is_on
+        self.is_on_lock.release()
+        if is_on:
+            self.is_face_detected = False
+            image = self.bridge.imgmsg_to_cv2(msg, desired_encoding="passthrough")
+            # Convert image to RGB colour
+            image_rgb = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
+            # Convert image to Grayscale
+            image_gray = cv2.cvtColor(image_rgb, cv2.COLOR_BGR2GRAY)
+
+            # Create an instance of the Face Detection Cascade Classifier
+            detector = cv2.CascadeClassifier(self.face_model_path)
+            # Detect faces using the haarcascade classifier on the grayscale image
+            faces = detector.detectMultiScale(image_gray)
+
+            # Check if a face has been detected
+            if len(faces) > 0:
+                self.is_face_detected = True
+
+            if self.is_face_detected:
+                # Create an instance of the Facial landmark Detector with the model
+                landmark_detector = cv2.face.createFacemarkLBF()
+                landmark_detector.loadModel(self.landmark_model_path)
+
+                # Detect landmarks on image
+                _, landmarks = landmark_detector.fit(image_gray, faces)
+                for landmark in landmarks:
+                    count = 0
+                    # Add a dummy face marker to the image
+                    cv_image = self.bridge.imgmsg_to_cv2(msg, "bgr8")
+                    for face in faces:
+                        # Draw a white coloured rectangle around each face
+                        (x, y, w, d) = face
+                        cv2.rectangle(
+                            cv_image, (x, y), (x + w, y + d), (255, 255, 255), 2
+                        )
+
+                # Display mouth landmarks (landmarks 48-67 are mouth landmarks, and the length of landmark[0] is 68)
+                for i in range(48, len(landmark[0])):
```

Comment:
You refer to `landmark` here, but this is outside the loop `for landmark in landmarks` which is confusing and seems rife for a bug.